### ISSUE TITLE
Data races in ContextWatcher

### ISSUE THREAD

**Issue by zikaeroh** [jackc/pgconn#25]
I'm seeing races like this in my CI after upgrading `pgconn`:

```
2020-01-12T06:29:03.4238312Z WARNING: DATA RACE
2020-01-12T06:29:03.4238463Z Write at 0x00c00025ecd9 by goroutine 70:
2020-01-12T06:29:03.4238768Z   github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch()
2020-01-12T06:29:03.4239018Z       /home/runner/go/pkg/mod/github.com/jackc/pgconn@v1.2.0/internal/ctxwatch/context_watcher.go:36 +0x63
2020-01-12T06:29:03.4239572Z   github.com/jackc/pgconn.(*PgConn).ayncClose.func1()
2020-01-12T06:29:03.4239784Z       /home/runner/go/pkg/mod/github.com/jackc/pgconn@v1.2.0/pgconn.go:526 +0x192
2020-01-12T06:29:03.4239968Z 
2020-01-12T06:29:03.4240321Z Previous read at 0x00c00025ecd9 by goroutine 10:
2020-01-12T06:29:03.4240978Z   github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Unwatch()
2020-01-12T06:29:03.4241429Z       /home/runner/go/pkg/mod/github.com/jackc/pgconn@v1.2.0/internal/ctxwatch/context_watcher.go:59 +0x9c
2020-01-12T06:29:03.4241734Z   github.com/jackc/pgconn.connect()
2020-01-12T06:29:03.4241987Z       /home/runner/go/pkg/mod/github.com/jackc/pgconn@v1.2.0/pgconn.go:251 +0x2552
2020-01-12T06:29:03.4242340Z   github.com/jackc/pgconn.ConnectConfig()
2020-01-12T06:29:03.4242560Z       /home/runner/go/pkg/mod/github.com/jackc/pgconn@v1.2.0/pgconn.go:137 +0x388
2020-01-12T06:29:03.4242734Z   github.com/jackc/pgx/v4.connect()
2020-01-12T06:29:03.4242947Z       /home/runner/go/pkg/mod/github.com/jackc/pgx/v4@v4.2.0/conn.go:198 +0x35d
2020-01-12T06:29:03.4243270Z   github.com/jackc/pgx/v4/stdlib.(*Driver).Open()
2020-01-12T06:29:03.4243659Z       /home/runner/go/pkg/mod/github.com/jackc/pgx/v4@v4.2.0/conn.go:105 +0x19e
2020-01-12T06:29:03.4243838Z   database/sql.(*dsnConnector).Connect()
2020-01-12T06:29:03.4244385Z       /opt/hostedtoolcache/go/1.13.6/x64/src/database/sql/sql.go:688 +0x88
2020-01-12T06:29:03.4244583Z   database/sql.(*DB).conn()
2020-01-12T06:29:03.4244832Z       /opt/hostedtoolcache/go/1.13.6/x64/src/database/sql/sql.go:1228 +0xa15
2020-01-12T06:29:03.4245031Z   database/sql.(*DB).PingContext()
2020-01-12T06:29:03.4245548Z       /opt/hostedtoolcache/go/1.13.6/x64/src/database/sql/sql.go:782 +0x92
2020-01-12T06:29:03.4245824Z   github.com/hortbot/hortbot/internal/pkg/docker/dpostgres.newDB.func1()
2020-01-12T06:29:03.4246084Z       /opt/hostedtoolcache/go/1.13.6/x64/src/database/sql/sql.go:800 +0x245
2020-01-12T06:29:03.4246325Z   github.com/hortbot/hortbot/internal/pkg/docker.(*Container).Start.func2()
2020-01-12T06:29:03.4246585Z       /home/runner/work/hortbot/hortbot/internal/pkg/docker/docker.go:62 +0x62
2020-01-12T06:29:03.4246803Z   github.com/cenkalti/backoff/v3.RetryNotify()
2020-01-12T06:29:03.4247064Z       /home/runner/go/pkg/mod/github.com/cenkalti/backoff/v3@v3.0.0/retry.go:37 +0xc8
2020-01-12T06:29:03.4247282Z   github.com/ory/dockertest/v3.(*Pool).Retry()
2020-01-12T06:29:03.4247746Z       /home/runner/go/pkg/mod/github.com/cenkalti/backoff/v3@v3.0.0/retry.go:24 +0xeb
2020-01-12T06:29:03.4247960Z   github.com/hortbot/hortbot/internal/pkg/docker.(*Container).Start()
2020-01-12T06:29:03.4248180Z       /home/runner/work/hortbot/hortbot/internal/pkg/docker/docker.go:61 +0x3d9
2020-01-12T06:29:03.4248391Z   github.com/hortbot/hortbot/internal/pkg/docker/dpostgres.newDB()
2020-01-12T06:29:03.4248623Z       /home/runner/work/hortbot/hortbot/internal/pkg/docker/dpostgres/dpostgres.go:42 +0x2eb
2020-01-12T06:29:03.4248840Z   github.com/hortbot/hortbot/internal/db/migrations_test.withDatabase()
2020-01-12T06:29:03.4249236Z       /home/runner/work/hortbot/hortbot/internal/pkg/docker/dpostgres/dpostgres.go:20 +0x5c
2020-01-12T06:29:03.4249462Z   github.com/hortbot/hortbot/internal/db/migrations_test.TestUpDown()
2020-01-12T06:29:03.4249690Z       /home/runner/work/hortbot/hortbot/internal/db/migrations/migrations_test.go:42 +0x52
2020-01-12T06:29:03.4249847Z   testing.tRunner()
2020-01-12T06:29:03.4250046Z       /opt/hostedtoolcache/go/1.13.6/x64/src/testing/testing.go:909 +0x199
2020-01-12T06:29:03.4250120Z 
```

Full logs for the build are shown here (including all race variants): https://github.com/hortbot/hortbot/commit/ba3d209a21a2df70997cf4f5486aaffe15a41b59/checks?check_suite_id=396092918

Looking at `ContextWatcher`, a number of field accesses are racy. The struct probably needs to be mutex'd, or refactored to work differently (the way it works seems perhaps a little awkward to my eye, but I haven't personally tried to fix this yet).

**Comment by zikaeroh** [comment]
As noted in #26, some of the new uses of `pgConn.contextWatcher` appear to be incorrect, namely those added in 98b3c57584a2bde785c3f706afcd3d371d6faec3.

**Comment by jackc** [comment]
> Looking at ContextWatcher, a number of field accesses are racy.

> As noted in #26, some of the new uses of pgConn.contextWatcher appear to be incorrect, namely those added in 98b3c57.

I think this is the fundamental issue. ContextWatcher was never designed to be concurrency safe, and the new usage for asynchronous close in 98b3c57 was incorrect to use pgConn.contextWatcher.

Can you try pgconn master?

**Comment by zikaeroh** [comment]
I just bumped my project to master, and I don't see the races anymore. Hopefully that means they're gone, but I'll watch.

That's enough for me, but I'm sure it'd be appreciated if you could tag some versions and propagate the bump it in pgx too so people aren't hanging. :slightly_smiling_face:

**Comment by davidsbond** [comment]
My team's project was also hit with this issue, we had dependabot raise a PR to upgrade from 1.1.0 to 1.2.0 this morning but the build failed due to the data race. Would be awesome if there could be a tagged version that has the race fixes so we can upgrade without fixing to master.

**Comment by jackc** [comment]
Just tagged pgconn v1.2.1 with this fix as well as pgx v4.2.1 which bumps its dependency on pgconn to v1.2.1.
